protocol/x/vault/keeper/orders.go

```diff
@@ -72,78 +73,46 @@ func (k Keeper) RefreshAllVaultOrders(ctx sdk.Context) {
+// last bit flipped only upon order replacement.
 func (k Keeper) RefreshVaultClobOrders(ctx sdk.Context, vaultId types.VaultId) (err error) {
-	// Cancel CLOB orders from last block.
-	orderIdsToCancel, err := k.GetVaultClobOrderIds(
-		ctx.WithBlockHeight(ctx.BlockHeight()-1),
-		vaultId,
-	)
-	if err != nil {
-		log.ErrorLogWithError(ctx, "Failed to get vault clob order IDs to cancel", err, "vaultId", vaultId)
-		return err
-	}
-	orderExpirationSeconds := k.GetParams(ctx).OrderExpirationSeconds
-	for _, orderId := range orderIdsToCancel {
-		if _, exists := k.clobKeeper.GetLongTermOrderPlacement(ctx, *orderId); exists {
-			err := k.clobKeeper.HandleMsgCancelOrder(ctx, clobtypes.NewMsgCancelOrderStateful(
-				*orderId,
-				uint32(ctx.BlockTime().Unix())+orderExpirationSeconds,
-			), true)
-			if err != nil {
-				log.ErrorLogWithError(ctx, "Failed to cancel order", err, "orderId", orderId, "vaultId", vaultId)
-			}
-			vaultId.IncrCounterWithLabels(
-				metrics.VaultCancelOrder,
-				metrics.GetLabelForBoolValue(metrics.Success, err == nil),
-			)
-		}
-	}
-	// Place new CLOB orders.
+	// Get client IDs of most recently placed orders, if any.
+	mostRecentClientIds := k.GetMostRecentClientIds(ctx, vaultId)
+	// Get orders to place.
 	ordersToPlace, err := k.GetVaultClobOrders(ctx, vaultId)
 	if err != nil {
 		log.ErrorLogWithError(ctx, "Failed to get vault clob orders to place", err, "vaultId", vaultId)
 		return err
 	}
 
-	for i, order := range ordersToPlace {
-		err := k.PlaceVaultClobOrder(ctx, order)
-		if err != nil {
-			log.ErrorLogWithError(ctx, "Failed to place order", err, "order", order, "vaultId", vaultId)
-		}
-
-		vaultId.IncrCounterWithLabels(
-			metrics.VaultPlaceOrder,
-			metrics.GetLabelForBoolValue(metrics.Success, err == nil),
-		)
-
-		// Send indexer messages. We expect orderIdsToCancel and ordersToPlace to have the same length
-		// and the order to place at each index to be a replacement of the order to cancel at the same index.
-		replacedOrderId := orderIdsToCancel[i]
-		if replacedOrderId == nil {
-			k.GetIndexerEventManager().AddTxnEvent(
-				ctx,
-				indexerevents.SubtypeStatefulOrder,
-				indexerevents.StatefulOrderEventVersion,
-				indexer_manager.GetBytes(
-					indexerevents.NewLongTermOrderPlacementEvent(
-						*order,
-					),
-				),
-			)
+	clientIds := make([]uint32, len(ordersToPlace))
+	for i, orderToPlace := range ordersToPlace {
+		if i >= len(mostRecentClientIds) { // when a vault first starts quoting or when `layers` increases.
+			// Place order.
+			err = k.PlaceVaultClobOrder(ctx, vaultId, orderToPlace)
 		} else {
-			k.GetIndexerEventManager().AddTxnEvent(
-				ctx,
-				indexerevents.SubtypeStatefulOrder,
-				indexerevents.StatefulOrderEventVersion,
-				indexer_manager.GetBytes(
-					indexerevents.NewLongTermOrderReplacementEvent(
-						*replacedOrderId,
-						*order,
-					),
-				),
-			)
```

Comment:
note: For expired orders, this breaks the order flickering logic since orders will have the expire + place events sent for them.

Reply:
correct. seems okay to me as expiration occurs much less frequently than before. 

however, one thing we can do might be to replace order (send replacement event) [here](https://github.com/dydxprotocol/v4-chain/blob/tq/tra-169/protocol/x/vault/keeper/orders.go#L99) (i.e. replace order if order gets filled or expires)